core/src/block_sync.rs

```diff
@@ -74,6 +82,23 @@ impl BlockSynchronizer {
 
     /// Sends request for latest blocks to a random peer
     async fn request_block(&mut self) {
+        let now_height = self.state.view().height() as u64;
+
+        if now_height > 0 {
+            let now_height = now_height.try_into().expect("Should not be zero");
+
+            // This guards against a softfork and adds general redundancy.
+            if now_height == self.latest_height {
+                self.seen_blocks.clear();
+            }
+            self.latest_height = now_height;
+
+            self.seen_blocks
+                .retain(|(height, _hash)| *height >= now_height);
+        } else {
+            self.seen_blocks.clear();
+        }
```

Comment:
I don't think you need to worry about the genesis

Reply:
Okay true.